scanpipe/pipes/scancode.py

```diff
@@ -310,18 +311,108 @@ def scan_for_files(project):
 
 def scan_for_application_packages(project):
     """
-    Runs a package scan on files without a status for a `project`.
+    Runs a package scan on files without a status for a `project`,
+    then create DiscoveredPackage and DiscoveredDependency instances
+    from the detected package data
 
     Multiprocessing is enabled by default on this pipe, the number of processes can be
     controlled through the SCANCODEIO_PROCESSES setting.
     """
     resource_qs = project.codebaseresources.no_status()
+
+    # Collect detected Package data and save it to the CodebaseResource it was
+    # detected from
     _scan_and_save(
         resource_qs=resource_qs,
         scan_func=scan_for_package_data,
         save_func=save_scan_package_results,
     )
 
+    # Iterate through CodebaseResources with Package data and handle them using
+    # the proper Package handler from packagedcode
+    assemble_packages(project=project)
+
+
+def add_to_package(package_uid, resource, project):
+    """
+    Relate a DiscoveredPackage to `resource` from `project` using `package_uid`
+    """
+    if not package_uid:
+        return
+    package_associated_with_resource = resource.discovered_packages.filter(
+        package_uid=package_uid
+    ).exists()
+    if not package_associated_with_resource:
+        package = project.discoveredpackages.get(package_uid=package_uid)
+        resource.discovered_packages.add(package)
+
+
+def assemble_packages(project):
+    """
+    Create instances of DiscoveredPackage and DiscoveredDependency for `project`
+    from the parsed package data present in the CodebaseResources of `project`.
+    """
+    seen_resource_paths = set()
+    package_data_resources_qs = project.codebaseresources.filter(
+        package_data__isnull=False
+    )
+    for resource in package_data_resources_qs:
+        if resource.path in seen_resource_paths:
+            continue
+
+        logger.info(
+            f"project: {project}:\n"
+            "function: assemble_packages\n"
+            f"Processing: CodebaseResource {resource.path}\n"
+        )
+
+        for package_data in resource.package_data:
```

Comment:
This wording is a bit confusing, we are using the same `package_data` but on the left side it's a single entry and on the right side it's a list.

Reply:
I've renamed the variables to avoid overloading `package_data`